### ISSUE TITLE
"runtime" uniform variables

### ISSUE THREAD

**Issue by wheaney** [ralgar/vkShade#14]
I thought you might be interested in [this vkBasalt PR](https://github.com/DadSchoorse/vkBasalt/pull/226) I just created. I suspect it will never get merged, so I'll probably be working off my own fork for now, but maybe you want to pull it in? Let me know and I can create a PR.

**Comment by ralgar** [comment]
Okay, I *really* like the potential of this idea - IPC opens up some very cool possibilities, and was something I intended to implement myself. The only thing that makes me hesitate with your implementation is that it's deviating from the official ReShadeFX language. The goal here is to unify as much as possible, so the last thing I want to do is cause a bunch of fragmentation where we now have Linux only "ReShadeFX" shaders.

On the other side of the problem, ReShade add-ons aren't exactly Linux compatible either, as they're DLLs that use the Win32 API for the main function, and they may potentially hook Win32 for other stuff as well.  Granted, it would be a relatively easy fix, using platform-specific preprocessor definitions, but I'm not sure whether add-on developers would care to do that or not. So, even when I eventually update the ReShade implementation (#6), implementing add-on support may not be terribly helpful in gaining this functionality or avoiding the fragmentation problem.

I'm thinking the right thing to do here (albeit a lot more work) is to create an actual API for this, allowing runtime modification of *any* loaded shader's uniforms through a dedicated Unix socket in XDG_RUNTIME_DIR. Correct me if I'm wrong, but that should accomplish what you wanted without deviating from the official implementation of the ReShadeFX language in any way.

**Comment by wheaney** [comment]
> deviating from the official ReShadeFX language

I'm new to ReShade FX but it does support runtime uniforms using the "runtime" source, the extra annotations would just get ignored. So technically (I *believe*) a shader like the one I wrote would be compatible with ReShade.

You're right that ReShade add-ons aren't Linux compatible, which is why I'm not using ReShade for this. I originally wrote my shader with an add-on and realized that since it would have to be compiled for Windows I wouldn't be able to hook into it with shared memory as easlily. So my uniform support is specifically *because* add-ons don't work, and is intended to by-pass that limitation. I'm sure add-ons do other things as well, but what I like about this PR is that it allows for updating uniform variables from an outside source *without* a big, complicated add-on needed.

I thought about sockets but that communication would be more complex. This was my thinking: my driver is sending out 1000 updates per second (reading IMU data from AR glasses), and the reshade shader is updating (frames) at probably 30-60 frames-per-second. If my driver is loading up the socket with data, it makes the vkBasalt implementation for reading from that socket much more complicated. What I like about shared memory is it's essentially broadcasting that information, it's not even necessarily for just this specific API. So if I have multiple things that I want to "hook" into my IMU data, I can write to shared mem once and multiple sources can read from it. It's also much simpler than an API integration would be, I don't like that ReShade add-ons are essentially plugins, so I have to compile some C/++ code against ReShade headers and go through the extra complication of that.

If all you need is to update some variables, this solution is so much simpler for both sides of the equation.

**Comment by ralgar** [comment]
>I'm new to ReShade FX but it does support runtime uniforms using the "runtime" source, the extra annotations would just get ignored. So technically (I believe) a shader like the one I wrote would be compatible with ReShade.

Oh, cool. Thanks for the clarification, I've not actually seen that feature in ReShadeFX before, but I don't exactly develop shaders for it myself. Do you happen to have any links to documentation on it, or other shaders (besides your own) that use it? I did a quick search, but didn't turn up much, and it would be helpful for me in learning more about this feature.

Your logic sounds sensible, and if it's only the annotations that are Linux-specific then it *shouldn't* be problematic. Based on what you're saying it sounds like it may also be possible to write shaders in a cross-platform manner just by including the annotations for both platforms?

**Comment by wheaney** [comment]
[Here's a comment](https://reshade.me/forum/addons-discussion/8442-providing-game-data-to-reshade#44602) I found from crosire talking about the feature. I'm glad you had me find this because my change is compatible with ReShade, but my approach to expecting all runtime uniforms to all use `source="runtime"` is not the best because anyone trying to use a shader (like the one I linked in the PR) in both ReShade and vkBasalt would become blocked by this if they wanted to have more than one runtime variable (since this `source` value is also the variable name in ReShade).

The fix to my PR would be easy. I have a couple options:
1. Update the vkBasalt uniform parsing code to expect all runtime uniforms to start with a prefix (e.g. `source="runtime:g_imu_pitch"`). This would be compatible with both ReShade and vkBasalt.
2. OR update the uniform parsing to assume any variable that doesn't match any of the other supported uniform types is a runtime variable. This is ideal because it matches exactly with how crosire described it in their comment.

These changes have the added benefit of making the annotations even simpler with optional `pathname` and `projId` annotations since the source values are unique. For example, if someone used `source="my_custom_var"` and didn't set a `pathname` or `projId` annotation, then vkBasalt would assume a path, something like `/tmp/my_custom_var`, and a projId of 0. Anyone wanting to have more control could use those optional annotations.

I really like (2) as the resulting shader FX code would be indistinguishable from a ReShade shader. I'll push an update to my PR containing what I've described in (2) here.
